app/lib/proxy_config_affecting_changes.rb

```diff
@@ -1,115 +1,158 @@
+      initial_state != object.reload.proxy_config_affecting_state.hash
+    rescue ActiveRecord::RecordNotFound
+      !new_record # Cannot find the model. If the object was a new record, then it was not persisted; otherwise, it was destroyed.
+    end
+
+    def changed?
+      destroyed? || created? || state_changed?
+    end
+
+    def proxies
+      @proxies.presence || find_proxies
+    end
+
+    protected
+
+    def find_object
+      object.class.find_by(id: object.id)
+    end
+
+    def find_proxies(model = find_object)
+      proxy_or_proxies = model.try(:proxy) || model.try(:proxies)
+      [*proxy_or_proxies].flatten.compact
     end
   end
 
-  module ProxyExtension
-    extend ActiveSupport::Concern
+  class Tracker
+    def initialize
+      @tracked_objects = []
+    end
 
-    included do
-      has_one :proxy_config_affecting_change, dependent: :delete
-      private :proxy_config_affecting_change
+    def track(object)
+      return unless object
+      tracked_object = TrackedObject.new(object)
+      return if tracking?(tracked_object)
+      @tracked_objects << tracked_object
+    end
 
-      include ProxyConfigAffectingChanges
+    def tracking?(object)
+      tracked_object = object.is_a?(TrackedObject) ? object : TrackedObject.new(object)
+      @tracked_objects.include?(tracked_object)
+    end
 
-      after_commit :issue_proxy_affecting_change_events, on: :update
+    def objects_with_affecting_changes
+      @tracked_objects.select(&:changed?)
+    end
 
-      def issue_proxy_affecting_change_events
-        return if previously_changed?(:created_at) || (previous_changes.keys - %w[updated_at lock_version]).empty?
+    def flush
+      proxies = objects_with_affecting_changes.map(&:proxies).flatten.uniq
+      proxies.each(&method(:issue_proxy_affecting_change_event))
+    ensure
+      @tracked_objects.clear
+    end
 
-        issue_proxy_affecting_change_event(self)
-      end
+    # FIXME: This is only so ProxyConfigs::AffectingObjectChangedEvent does not crash
+    def id
+      Thread.current.name
+    end
 
-      def find_or_create_proxy_config_affecting_change
-        proxy_config_affecting_change || create_proxy_config_affecting_change
-      end
-      alias affecting_change_history find_or_create_proxy_config_affecting_change
-      private :find_or_create_proxy_config_affecting_change
-
-      def pending_affecting_changes?
-        return unless apicast_configuration_driven?
-        config = proxy_configs.sandbox.newest_first.first
-        return false unless config
-        config.created_at < affecting_change_history.updated_at
-      end
+    protected
 
-      private
+    def issue_proxy_affecting_change_event(proxy)
+      # Sometimes invoked in the context of or concurrently to a hierarchy deletion,
+      # hence the proxy, the service or the account may no longer be available
+      return unless proxy
+      service = proxy.service
+      return unless service&.account
```

Comment:
If we are using the `&.` operator why not chaining directly as you do not need `service` nor `account` variable

```ruby
return unless proxy&.service&.account
```